src/v/raft/replicate_batcher.cc

```diff
@@ -72,14 +72,25 @@ replicate_batcher::cache_and_wait_for_result(
          * replicate batcher stop method
          *
          */
-        ssx::background
-          = ssx::spawn_with_gate_then(_bg, [this]() {
-                return _lock.get_units().then([this](auto units) {
-                    return flush(std::move(units), false);
-                });
-            }).handle_exception([item](const std::exception_ptr& e) {
-                item->set_exception(e);
+        if (!_flush_pending) {
+            _flush_pending = true;
+            ssx::background = ssx::spawn_with_gate_then(_bg, [this]() {
+                return _lock.get_units()
+                  .then([this](auto units) {
+                      return flush(std::move(units), false);
+                  })
+                  .handle_exception([this](const std::exception_ptr& e) {
+                      // an exception here is quite unlikely, since the flush()
+                      // method generally catches all its exceptions and
+                      // propagates them to the promises associated with the
+                      // items being flushed
+                      vlog(
```

Comment:
@mmaslankaprv I think this is the main possibly controversial part of this change. The existing code propagates exceptional futures from flush() to the just-cached item. This doesn't really make too much sense with the new code, because often no flush will even be enqueued (when `flush_pending` is already true): some "other" flush will flush the item. This was also true before, thought less obvious from the code: there is no guarantee which flush actually flushes the item and in the case of the OOM the flushes were delayed by hours from when they were issued. So should a flush actually fail it would often be the case that `->set_exception` occurs on a promise that has already been satisfied by an earlier flush.

I checked the behavior of flush and `do_flush` which it calls and for `flush` it never throws or returns an exceptional future: rather exceptions are propagated to the items it is trying to flush. `do_flush` is mostly the same but there seems to be some code near the end not protected by a `try` (e.g., the `_bg.is_closed()` case) though it is not clear that it ever throws.

So the current change just logs these errors, but I could also restore the old behavior.

Reply:
this looks good, i am wondering if we would be able to achieve the same semantics by checking the size of waiters queue i.e. if there are waiters skip flush. ( think having `bool` is better as the intent is better communicated, but was just wonering)

Reply:
> i am wondering if we would be able to achieve the same semantics by checking the size of waiters queue i.e. if there are waiters skip flush.

Hmm, which queue did you have in mind? The only one I can see is the queue on the `_lock` mutex, though the waiter does not appear immediately when a flush is enqueued: first the background task has to run, then the `_lock` must be contended so it will wait on the mutex, then another fiber could observe the waiter. I think it would still work because as soon as you accumulate a little backlog of flushes the `_lock` will almost constantly have one waiter, but at least initially it will cause more flushes than necessary. 

As you say, having it explicit is a bit clearer, but it would also be nice to wrap this up in some pattern like `async_singleton_job` or something, where anyone could notify the job and it would ensure that after every notification the job would run at least once, but would avoid buliding up any backlog of jobs.

Reply:
i was thinking about the `_lock` waiters queue, i thought when creating a background task there is no scheduling point involved till the first `co_await`?

Reply:
Hmm, I see, so you are saying that,given the following:

```
out1();
ssx::spawn_with_gate_then(_bg, [this]() {
                inside();
                return _lock.get_units().then([this](auto units) {
                    return flush(std::move(units), false);
                });
            })
out2();
```

`inside()` will usually run synchronously on the same fiber after `out1` (obviously) but before `out2`, right?

Good point. Every `then` is a potential scheduling point though if the future is called on is ready then I believe it executes immediately without creating a new task "usually" if the quota hasn't been exceeded and also some counter of 255 synchronous executions on the current fiber hasn't been exceeded. 

So maybe you are correct that immediately after running this, the `_lock` will already have 1 waiter or perhaps is held already by the fiber which was able to enter it (and eventually blocked).